inputremapper/gui/data_bus.py

```diff
@@ -0,0 +1,171 @@
+import os.path
+import re
+import traceback
+from collections import defaultdict, deque, namedtuple
+from dataclasses import dataclass
+from enum import Enum
+from typing import (
+    Callable,
+    Dict,
+    Set,
+    Protocol,
+    Tuple,
+    Deque,
+    Optional,
+    List,
+    NamedTuple,
+)
+
+from inputremapper.event_combination import EventCombination
+from inputremapper.logger import logger
+
+
+class MessageType(Enum):
+    reset_gui = "reset_gui"
+    terminate = "terminate"
+    init = "init"
+
+    uinputs = "uinputs"
+    groups = "groups"
+    group = "group"
+    preset = "preset"
+    mapping = "mapping"
+    combination_recorded = "combination_recorded"
+    recording_finished = "recording_finished"
+    combination_update = "combination_update"
+    status = "status"
+
+    # for unit tests:
+    test1 = "test1"
+    test2 = "test2"
+
+
+class Message(Protocol):
+    """the protocol any message must follow to be sent with the DataBus"""
+
+    message_type: MessageType
+
+
+# useful type aliases
+MessageListener = Callable[[Message], None]
+Capabilities = Dict[int, List]
+Name = str
+Key = str
+DeviceTypes = List[str]
+
+
+class DataBus:
```

Comment:
In which scenario would _messages contain more than one message?

I know this kind of service as a MessageBroker, like mosquitto or kafka

Reply:
when multiple listeners send messages they will be added to the `_messages` queue and sent once the current message is sent to the listeners.

Reply:
but every `send` calls `_send_all` right away

Reply:
and `_send_all` returns immediately if it is already running. See `TestDataBus.test_preserves_order` this test will queue up 3 messages.

The idea is that all messages are always sent in the order in which they arrive. For example `Controller.on_preset_changed()` listens for a `preset` message and immediately responds by sending a `mapping` message. But we don't want this message to be sent immediately, first the `preset` message must be distributed to all listeners. 
Otherwise a listener who listens to both the `preset` and the `mapping` might receive the `mapping` first and mess up something.

Sure, we can rename it to message broker if this is a more common name.

Reply:
Are multiple threads interacting with the DataBus?

Reply:
No.

Reply:
While the loop below is running, how would the python process ever have a chance of calling `send`?

since `_send` is not async and there are no other threads, the process will block until the loop is done.

Ah, except if the listener calls `send` as well. A second listener would then get the two messages in reverse order.

Reply:
> Ah, except if the listener calls `send` as well. A second listener would then get the two messages in reverse order.

exactly `TestDataBus.test_preserves_order` demonstrates this